verification/spec/connection-handshake/ConnectionHandshakeModule.tla

```diff
@@ -0,0 +1,353 @@
+    reported in the client proof must not exceed the current (latestHeight) of
+    this chain.
+ *)
+VerifyClientProof(cp) ==
+    cp.height <= store.latestHeight
+
+
+(***************************************************************************
+ Connection Handshake Module actions & operators.
+ ***************************************************************************)
+
+
+(* Modifies the local store.
+
+    Replaces the connection in the store with the argument 'newCon'.
+    This action also advances the chain height. 
+ *)
+NewStore(newCon) ==
+    [store EXCEPT !.connection = newCon,
+                  !.latestHeight = @ + 1]
+
+
+(* Handles a "ICS3MsgInit" message 'm'.
+    
+    Primes the store.connection to become initialized with the parameters
+    specified in 'm'. Also creates a reply message, enqueued on the outgoing
+    buffer.
+ *)
+HandleInitMsg(m) ==
+    LET newCon == [parameters |-> m.parameters,
+                   state      |-> "INIT"]
+        sProof == GetConnProof("INIT")
+        cProof == GetClientProof
+        replyMsg == [parameters |-> FlipConnectionParameters(m.parameters),
+                     type |-> "ICS3MsgTry",
+                     connProof |-> sProof,
+                     clientProof |-> cProof] IN
+    IF /\ ValidConnectionParameters(m.parameters)
+       /\ store.connection.state = "UNINIT"
+    THEN [out |-> Append(outBuf, replyMsg),
+          store |-> NewStore(newCon)]
+    ELSE [out |-> outBuf,
+          store |-> store]
+
+
+(* State predicate, guarding the handler for the Try msg.
+
+    If any of these preconditions does not hold, the message
+    is dropped. 
+ *)
+PreconditionsTryMsg(m) ==
+    /\ ValidConnectionParameters(m.parameters)
+    /\ \/ store.connection.state = "UNINIT"
+       \/ /\ store.connection.state = "INIT"
+          /\ CheckLocalParameters(m.parameters)
+    /\ m.connProof.connectionState = "INIT"
+    /\ VerifyConnProof(m.connProof)
+    /\ VerifyClientProof(m.clientProof)
+
+
+(* Handles a "ICS3MsgTry" message.
+ *)
+HandleTryMsg(m) ==
+    LET newCon == [parameters |-> m.parameters, 
+                   state |-> "TRYOPEN"]
+        sProof == GetConnProof("TRYOPEN")
+        cProof == GetClientProof
+        replyMsg == [parameters |-> FlipConnectionParameters(m.parameters),
+                     type |-> "ICS3MsgAck",
+                     connProof |-> sProof,
+                     clientProof |-> cProof] IN
+    IF PreconditionsTryMsg(m)
+    THEN [out |-> Append(outBuf, replyMsg),
+          store |-> NewStore(newCon)]
+    ELSE [out |-> outBuf,
+          store |-> store]
+
+
+(* State predicate, guarding the handler for the Ack msg. 
+ *)
+PreconditionsAckMsg(m) ==
+    /\ \/ store.connection.state = "INIT"
+       \/ store.connection.state = "TRYOPEN"
+    /\ CheckLocalParameters(m.parameters)
+    /\ m.connProof.connectionState = "TRYOPEN"
+    /\ VerifyConnProof(m.connProof)
```

Comment:
I think we also need `VerifyClientProof(m.clientProof)` here?

Reply:
I am not sure if it's a good idea to separate preconditions in a separate predicate.